src/main/java/com/exadel/discountwebapp/category/validator/CategoryValidator.java

```diff
@@ -17,10 +23,35 @@ public void validate(CategoryRequestVO request) {
         checkDuplicateTitle(request);
     }
 
-   public void checkDuplicateTitle(CategoryRequestVO request) {
+    public void checkDuplicateTitle(CategoryRequestVO request) {
         categoryRepository.findByTitle(request.getTitle())
                 .ifPresent(category -> {
                     throw new EntityAlreadyExistsException(Category.class, "title", category.getTitle());
                 });
     }
+
+    public void checkDuplicateTag(Category category, List<Tag> newTags) {
+        newTags.forEach(e -> {
+            if (category.getTags().contains(e)) {
+                throw new EntityAlreadyExistsException(Tag.class, "name", e.getName());
+            }
+        });
+    }
+
+    public void checkTagAlreadyUsedInDiscount(Category category, List<Long> tagIds) {
+        List<Discount> discounts = category.getDiscounts();
```

Comment:
Here we are retrieving all discounts of the category.
The amount of data retrieved from the database can be too large.
Probably this check can be performed via a single SQL query without retrieving all the data from the database.
But let's make this optimization later.